File: 238-ProductOfArrayExceptSelf/238-ProductOfArrayExceptSelf.cpp
// Last updated: 7/29/2026, 10:13:20 AM
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        std::vector<int> sln(nums.size(), 1);
        int prefix = 1;
        int postfix = 1;
        // prefix loop
        for(int i = 0; i < nums.size(); ++i){
                sln[i] = prefix;
                prefix *= nums[i];
            }
        // postfix loop
        for(int i = nums.size()-1; i >= 0 ; --i){
                sln[i] *= postfix;
                postfix *= nums[i];
            }
            return sln;

    }
};

File: 238-ProductofArrayExceptSelf/238-ProductofArrayExceptSelf.py
# Last updated: 8/1/2026, 9:12:39 AM
1class Solution:
2    def productExceptSelf(self, nums: List[int]) -> List[int]:
3        # 2 passes through the nums array, the first stores prefix in answer, the second multiplies the answers by post fix
4        answer = [1] * (len(nums))
5        prefix = 1
6        postfix = 1
7
8        for i in range(len(nums)):
9            answer[i] = prefix
10            prefix = prefix * nums[i]
11        for i in range(len(nums) - 1, -1, -1):
12            answer[i] *= postfix
13            postfix = postfix * nums[i]
14        return answer